### ISSUE TITLE
Automatic reference import and formatting support?

### ISSUE THREAD

**Issue by and3k** [asciidoctor/asciidoctor#1374]
Hi!

AsciiDoctor is really great, but to my knowledge there is no way to automatically import cited references from a file (i.e. BibTeX) and format them properly (accord to the given CSL). This is something that pandoc does very well (http://pandoc.org/README.html#citations).

In the past I used AsciiDoc to write scientific articles and used Zotero to imported formatted references, but that gets annoying really fast and so now I’m mostly using markdown. IMHO this is a major blocker for get the scientific community to use AsciiDoc and finally stop using Word :)

Not sure if this in the scope of AsciiDoc(tor) but it would be great to have!

Thanks!
Bela

**Comment by jxxcarlson** [comment]
Send me typical bibtex file at jxxcarlson@gmail.com <mailto:jxxcarlson@gmail.com>
and I will see what I can do.

Cheers!
JIm

**Comment by and3k** [comment]
Done. Thanks!

**Comment by shahryareiv** [comment]
I use asciidoc-bib to process the text before processing by asciidoctor but it has limitations (and bugs). Furthermore this should be a part of core asciidoctor. Asciidoc-bib and some other similar solutions are based on citeproc (citeproc-ruby) and there should be a natural way to integrate it . Asciidoc also uses a special format [cite:X] which I guess should be cite:[X] as it looks like a macro (?). Features such as bibliography at the end of chapters, at footnotes, at sidebars (Edward Tufte style) are also other features not available (in backends probably).

**Comment by and3k** [comment]
I didn’t know about asciidoc-bib. But as you already pointed out there are some limitations. One I already see is it’s inability to automatically add "et al." when appropriate. I like the possibility to add page numbers (can’t be done with pandoc afaik), but I also wouldn’t know when to use this when writing a scientific paper ;) 

Btw, bibliography at the end of chapters is something that can be done with the DocBook style sheets (I think).

**Comment by shahryareiv** [comment]
The format of citation in asciidoc-bib is [cite:X;Y] which I think cite:[X,Y] is more compatible semantically with asciidoc. Although it is based on CiteProc but I have no idea how to apply a specific CLS. Placement of the bibliography should also be determined in the text not in the backend configuration (I think). I also discovered asciidoc-bib has problem with dashes in the bibkey.

**Comment by mojavelinux** [comment]
> The format of citation in asciidoc-bib is [cite:X;Y] which I think cite:[X,Y] is more compatible semantically with asciidoc.

100% agreed. This is one of the hangups I have with integrating asciidoc-bib into Asciidoctor. I'd like to see that changed...perhaps an option that can be passed to asciidoc-bib to set which syntax is followed.

**Comment by mojavelinux** [comment]
Related thread: http://discuss.asciidoctor.org/Q-Generated-per-doc-bibliography-from-adoc-bibliography-master-files-td3519.html

**Comment by jxxcarlson** [comment]
Agree totally!

**Comment by jirutka** [comment]
Support for BibTeX would be indeed useful!

However I disagree that it should be integrated directly into the core. In my opinion, core should be small as possible and all extra features should be in extensions (separate gems).

**Comment by shahryareiv** [comment]
Bibtex is just one format from many bibliographic data formats (BibLatex, RIS, RefWors, ..). Also Bibtex is combined with ".bst" files for visual formatting bibliographies. Asciidoc-bib uses Citeproc library, which respectively uses CSL formatting (http://citationstyles.org/). I think Citeproc is a good choice, regarding its extensive support (and probably functionality, over 7500 styles!). What remains is actually collecting and locating citations and trying to figure out where to put the bibiliographies (end of chapter, end of book, both, sidenote, footnote, ...) hence I think this is partially a job of asciidoctor core.

**Comment by and3k** [comment]
I agree that citeproc (i.e., CSL) is the way to go.

And I think the most important feature should be only showing references which are actually cited in the text (or explicitly added, see `nocite` in pandoc-citeproc) and not all the references in a bib file.

**Comment by nightscape** [comment]
Just wanted to note that there is a fork of `asciidoctor-bib` called [`asciidoctor-bibtex`](https://github.com/ProgramFan/asciidoctor-bibtex) which changes the syntax to the preferred `cite[...]` style and supports more (I think even all) options of asciidoctor itself.
Btw, could reference management not be done by a pure extension instead of a "precompiler"?

**Comment by mojavelinux** [comment]
Yes, I would prefer if these were done using macro extensions instead of a preprocessor. In order for a preprocessor to respect the AsciiDoc syntax fully, it has to reimplement the AsciiDoc parser. That's why the macro extensions are better because you can build on the parser instead of sidestepping it.

**Comment by seinecle** [comment]
Hello,
Reading this thread now. Wondering if the work done by Zotero could be built upon here? They have extensions for MS Word, maybe can they provide the resources to build the equivalent for asciidoctor? I ping them on this issue.

**Comment by shahryareiv** [comment]
@seinecle That integration is editor-software based, asciidoc is editor-software neutral. I have used zotero with asciidoc, simply by copy pasting bibtex keys in the citation format of asciidoc and using asciidoc-bib (no more) or asciidoctor-bibtex extensions (only for HTML output). Both asciidoc-bib and asciidoctor-bibtex are based on CiteProc and CSL where Zotero is one the of main supporters (http://citationstyles.org/). I feel that  asciidoctor-bibtex  needs some more enhancements but generally it is in production-level.

One suggestion can be to ask Zotero or Zotero Better Bibtex (https://github.com/retorquere/zotero-better-bibtex) to add support for asciidoc format copy-paste. Beter Bibtex has a record of facilitating integration of different editors with Zotero https://github.com/retorquere/zotero-better-bibtex/wiki/Cite-as-you-Write. Anyway that saves only a few key presses in each copy-paste and is not very crucial.

**Comment by ronaldtse** [comment]
cc: @mojavelinux  Some time ago we created a new called [`asciidoctor-bibilography`](https://github.com/riboseinc/asciidoctor-bibliography) that aims to solve citation issues encountered using asciidoctor. We have had a strong need for this ourselves especially when converting TeX documents to AsciiDoc.

This gem fully bridges CSL styles with a syntax designed to be "native-asciidoctor". It also supports all TeX-style citation styles like `citep:[]`, `citet:[]` for TeX-lovers. 😉 

Please feel free to provide feedback if anyone encounter any issues or has any further wants from the gem. Thanks! 😄

**Comment by Fiech** [comment]
@ronaldtse Just wanted to say that I just found your gem and am very happy with how it works. Great work! I'm just in the process of trying stuff with my book publishing process and already have put in a pull request.
